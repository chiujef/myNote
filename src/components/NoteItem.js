import PropTypes from 'prop-types';
import React from 'react';
import {StyleSheet, View} from 'react-native';
import {Button, CheckBox} from 'react-native-elements';
import {FocusableTextInput} from '@components/core';

const NoteItem = ({
  noteItem,
  isSelected,
  onCompletedToggle,
  onChangeText,
  onTextPress,
  onDeletePress,
}) => {
  function onThisCompletedToggle() {
    onCompletedToggle && onCompletedToggle(noteItem.id);
  }

  function onThisChangeText(text) {
    onChangeText && onChangeText(noteItem.id, text);
  }

  function onThisDeletePress() {
    onDeletePress && onDeletePress(noteItem.id);
  }

  function onThisTextPress() {
    onTextPress && onTextPress(noteItem.id);
  }

  return (
    <View style={styles.noteItemContainer}>
      <CheckBox
        style={styles.noteItemCheckBox}
        checked={noteItem.completed}
        onPress={onThisCompletedToggle}
      />
      <FocusableTextInput
        style={styles.noteItem}
        textInputStyle={noteItem.completed ? styles.noteItemCompleted : null}
        text={noteItem.text}
        isFocus={isSelected}
        onChangeText={onThisChangeText}
        onPress={onThisTextPress}
      />
      <View style={styles.noteItemButtonContainer}>
        {isSelected ? <Button title="x" onPress={onThisDeletePress} /> : false}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  noteItemContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-start',
  },
  noteItemCheckBox: {
    flex: 0.15,
  },
  noteItem: {
    marginTop: 8,
    fontSize: 18,
    fontWeight: '400',
    flex: 0.6,
  },
  noteItemCompleted: {
    textDecorationLine: 'line-through',
    textDecorationStyle: 'solid',
  },
  noteItemButtonContainer: {
    alignItems: 'flex-start',
    justifyContent: 'center',
    margin: 1,
    flex: 0.25,
  },
  noteText: {
    textAlignVertical: 'center',
  },
});

NoteItem.propTypes = {
  noteItem: PropTypes.shape({
    id: PropTypes.number,
    text: PropTypes.string,
    completed: PropTypes.bool,
  }),
  isSelected: PropTypes.bool,
  onCompletedToggle: PropTypes.func,
  onChangeText: PropTypes.func,
  onTextPress: PropTypes.func,
  onDeletePress: PropTypes.func,
};

export default NoteItem;
