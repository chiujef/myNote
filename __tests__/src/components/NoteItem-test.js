/**
 * @format
 */

import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {render, fireEvent} from 'react-native-testing-library';

import NoteItem from '@components/NoteItem';

describe('NoteItem', () => {
  let noteItem;
  let noteItemComplete;

  beforeEach(() => {
    noteItem = {
      id: 1,
      text: 'tent',
      completed: false,
    };

    noteItemComplete = {
      id: 1,
      text: 'tent',
      completed: true,
    };
  });

  describe('Rendering', () => {
    it('should match to snapshot', () => {
      const component = renderer
        .create(<NoteItem noteItem={noteItem} isSelected={true} />)
        .toJSON();
      expect(component).toMatchSnapshot();
    });

    it('completed noteItem should match to snapshot', () => {
      const component = renderer
        .create(<NoteItem noteItem={noteItemComplete} isSelected={true} />)
        .toJSON();
      expect(component).toMatchSnapshot();
    });

    it('isSelected should show "x" button', () => {
      const {queryByText} = render(
        <NoteItem noteItem={noteItem} isSelected={true} />,
      );

      expect(queryByText('x')).not.toBeNull();
    });

    it('not isSelected should not show "x" button', () => {
      const {queryByText} = render(
        <NoteItem noteItem={noteItem} isSelected={false} />,
      );

      expect(queryByText('x')).toBeNull();
    });
  });

  describe('Pressing checkbox', () => {
    it('should trigger onCompletedToggle event with noteItem id', () => {
      const onCompletedToggleHandler = jest.fn();
      const {getByTestId} = render(
        <NoteItem
          noteItem={noteItem}
          onCompletedToggle={onCompletedToggleHandler}
        />,
      );

      fireEvent(getByTestId('checkbox'), 'onPress');

      expect(onCompletedToggleHandler).toHaveBeenCalledWith(noteItem.id);
    });
  });

  describe('Pressing unselected noteItem text', () => {
    it('should trigger onTextPress with noteItem id', () => {
      const onTextPressHandler = jest.fn();
      const {queryByText} = render(
        <NoteItem noteItem={noteItem} onTextPress={onTextPressHandler} />,
      );

      fireEvent(queryByText('tent'), 'onPress');

      expect(onTextPressHandler).toHaveBeenCalledWith(noteItem.id);
    });
  });

  describe('Changing selected noteItem text', () => {
    it('should trigger onChangeText with noteItem id', () => {
      const onChangeTextHandler = jest.fn();
      const {queryByType} = render(
        <NoteItem
          noteItem={noteItem}
          isSelected={true}
          onChangeText={onChangeTextHandler}
        />,
      );

      fireEvent.changeText(queryByType('TextInput'), 'ab');

      expect(onChangeTextHandler).toHaveBeenCalledWith(noteItem.id, 'ab');
    });
  });

  describe('Pressing "x" button', () => {
    it('should trigger onThisDeletePress event with noteItem id', () => {
      const onDeletePressHandler = jest.fn();
      const {getByText} = render(
        <NoteItem
          noteItem={noteItem}
          isSelected={true}
          onDeletePress={onDeletePressHandler}
        />,
      );

      fireEvent(getByText('x'), 'onPress');

      expect(onDeletePressHandler).toHaveBeenCalledWith(noteItem.id);
    });
  });
});
