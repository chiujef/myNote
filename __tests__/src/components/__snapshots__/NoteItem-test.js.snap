// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteItem Rendering completed noteItem should match to snapshot 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "flex-start",
    }
  }
>
  <View
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "margin": 5,
        "marginLeft": 10,
        "marginRight": 10,
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="checkbox"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#2089dc",
              "fontSize": 24,
            },
            Object {
              "minWidth": 24,
            },
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCompleteType="off"
      autoCorrect={false}
      autoFocus={true}
      blurOnSubmit={false}
      multiline={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      rejectResponderTermination={true}
      style={
        Array [
          undefined,
          Object {
            "textDecorationLine": "line-through",
            "textDecorationStyle": "solid",
          },
        ]
      }
      underlineColorAndroid="transparent"
      value="tent"
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flex": 0.25,
        "justifyContent": "center",
        "margin": 1,
      }
    }
  >
    <View
      style={
        Object {
          "borderRadius": 3,
        }
      }
    >
      <View
        accessible={true}
        clickable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2089dc",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingBottom": 1,
                "paddingTop": 2,
                "textAlign": "center",
              }
            }
          >
            x
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`NoteItem Rendering should match to snapshot 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "flex-start",
    }
  }
>
  <View
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "margin": 5,
        "marginLeft": 10,
        "marginRight": 10,
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="checkbox"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#bfbfbf",
              "fontSize": 24,
            },
            Object {
              "minWidth": 24,
            },
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCompleteType="off"
      autoCorrect={false}
      autoFocus={true}
      blurOnSubmit={false}
      multiline={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      rejectResponderTermination={true}
      style={
        Array [
          undefined,
          null,
        ]
      }
      underlineColorAndroid="transparent"
      value="tent"
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flex": 0.25,
        "justifyContent": "center",
        "margin": 1,
      }
    }
  >
    <View
      style={
        Object {
          "borderRadius": 3,
        }
      }
    >
      <View
        accessible={true}
        clickable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2089dc",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingBottom": 1,
                "paddingTop": 2,
                "textAlign": "center",
              }
            }
          >
            x
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
